draft-ietf-taps-interface.md

```diff
@@ -1436,6 +1451,25 @@ stack supports it, there is no guarantee that a soft error will be signaled.
 Connection -> SoftError<>
 ~~~
 
+# Ordering of Operations and Events
+
+As this interface is designed to be independent of concurrency model, the
+details of how exactly actions are handled, and on which threads/callbacks
+events are dispatched, is an implementation detail. However, the interface
+does provide the following guarantees about the ordering of operations:
+
+- Received<> will never occur on a Connection before a Ready<> event on that
+  Connection, or a ConnectionReceived<> or RendezvousDone<> containing that
+  Connection.
+
+- No events or errors will occur on a Connection after a Closed<> event, an
+  InitiateError<> or ConnectionError<> on that connection. To ensure this
+  ordering, neither Closed<> nor ConnectionError<> will occur on a Connection
+  while other events on the Connection are still outstanding.
```

Comment:
How can you guarantee that?  An abort can happen at any time, and that would give you a ConnectionError. Maybe I don't understand what "outstanding" means... or maybe it needs to be defined?

Reply:
This really depends on the concurrency model of the implementation platform. If the platform does callbacks, it ensures this by enqueueing dispatch of close/connection error after the dispatch of all other events is completed. If the platform does CSP dispatch, this naturally happens because a channel creates a queue. If the platform dispatches every event to a new thread, then this is hard, but it's kind of the platform's fault then...

Reply:
Okay, you mean "outstanding" as in "enqueued locally". Then this should be phrased as "locally outstanding" or something like that, I think.  To be clear, I have no problem at all with the first sentence - it's the second one that makes me think of the protocol machinery when I read "outstanding" - as in outstanding packets, incoming messages, causing events...

Reply:
Does this mean `Closed<>` can't fire until you're sure there can be no errors from the peer (e.g., after the TIME_WAIT timeout for TCP)?

Reply:
I agree with the approach of always having some final event (whether this is Closed or something else like Invalidated is another point).

If the connection has been dropped by the application, then I think the assumption is that if there are any errors that occurred later, the application would have no way of receiving them anyhow, so they must be silently discarded?

Reply:
+1 to "locally outstanding", will fix. Will have to think about `Closed<>` versus `ConnectionError<>`.